README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# SRS

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![](https://img.shields.io/github/last-commit/QianYe-stat/SRS.svg)](https://github.com/QianYe-stat/SRS/commits/master)
[![GitHub
issues](https://img.shields.io/github/issues/QianYe-stat/SRS)](https://github.com/QianYe-stat/SRS/issues)
[![License:
MIT](https://img.shields.io/badge/license-MIT-blue.svg)](https://cran.r-project.org/web/licenses/MIT)

<!-- badges: end -->

The goal of SRS is to to take one or multiple simple random samples
(SRS) of the specified size from an assumed population which are
elements of a vector with length\>1, and return the summary statistics
including sample mean, sample standard deviation and etc.

The functions in SRS can additionally be used to illustrate the
asymptotic properties of estimates based on SRS, which is also known as
the central limit theorem.

## Installation

You can install SRS from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("QianYe-stat/SRS")
```

## Code of Conduct

Please note that the SRS project is released with a [Contributor Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.

## Development of SRS

  - Initialize the package folder using
    `create_package("~/Mye.package/SRS")`
  - Initialize Git repo using `use_git()`.
  - Use `use_r("is_wholenumber")` and `use_r("SRSampling")` to create 2
    separate blank R script files. Copy and paste the codes for the 2
    functions from assignment-1b to the 2 R script files respectively.
  - Commit changes and run `check()`
  - Modify `DESCRIPTION`. Mostly manually but with
    `use_mit_license("Qian Ye")` to fill-in the license information.
  - Run `use_package("stats")` to add `stats` package to Import field in
    `DESCRIPTION` as `sd()` function from `stats` is used in my package.
  - Update my R codes with `sd` replaced by `stats::sd`.
  - Run `check()`, `install()` and then `library(SRS)`. It’s working\!
  - Create `tests` folder using `use_testthat()`
  - Create a test file using `use_test("SRSampling")` and write
    `test_that` functions to test `SRSampling` function. Likewise create
    tests for `is_wholenumber`.
  - Run `test()`. Passed\!
  - Update R codes with a specially formatted comment begin with `#'`.
    Run `document()` to generate corresponding `*.Rd` files. Try
    `?SRSampling`. It’s working\!
      - note that, I choose not to `@export` for the `is_wholenumber`
        function. Because it’s sourced from `examples` under
        `integer{base}` and out of the objective of `SRS` package which
        is to take SRS samples.
  - Manually configured a GitHub Personal Access Token (PAT) following
    the
    [instructions](https://happygitwithr.com/github-pat.html#github-pat).
  - Run `use_github()` to connect the local `SRS` package and Git
    repository to a companion repository on GitHub.
  - Create `README` file by using `use_readme_rmd()`.
  - Manually modify the `README.Rmd` file and run `build_readme()` to
    render it.
  - Create a vignette file by using
    `use_vignette("Use_SRS_to_Show_CLT")`.
  - Manually modify it and run `build_vignettes()`
  - Add a code of conduct by using `use_code_of_conduct()`, and describe
    the code of conduct in README.
  - Make a package website:
      - first install `pkgdown` package, and then run `use_pkgdown()`
        and `pkgdown::build_site()`;
      - create Github Pages based on `/docs`;
      - Please see the package website
        [here](https://qianye-stat.github.io/SRS/)
  - Create the `NEWS.md` file using `use_news_md()` and modify it.
  - Use `badger` to add badges.
  - Final `check()` and `install()`.

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(SRS)
## basic example code
###### informal tests #######
pop.1 <- rnorm(1000,0,10)

# take a SRS from x of size=10, show.SRS=TRUE
SRSampling(x=pop.1, size=10, show.SRS=T)
#> A simple random sample of size 10 is drawn from x
#> 
#> The SRS for each run of samplings are shown in the result
#> $sample.mean
#> [1] -3.039676
#> 
#> $sample.se
#> [1] 4.912975
#> 
#> $SRS
#>           [,1]      [,2]      [,3]      [,4]       [,5]     [,6]      [,7]
#> [1,] -5.006548 -10.07339 -10.35358 0.3601517 -0.7644633 3.230316 -5.820861
#>          [,8]      [,9]     [,10]
#> [1,] 3.436945 -1.212003 -4.193337

# take 20 SRS from x of size=10, show.SRS=FALSE
SRSampling(x=pop.1, size=10, rep=20)
#> 20 runs of simple random sample of size 10 are drawn from x
#> 
#> The SRS for each run of samplings are NOT included in the result
#> $sample.mean
#> $sample.mean$values
#>  [1] -2.8360801 -1.1082398 -8.3983181 -5.0031040  0.4403522 -3.4828091
#>  [7] -1.1641312  1.1262584 -2.4292281  3.8162037 -7.8090442  0.3320144
#> [13]  0.4622417 -1.7011468 -1.8747509 -1.7828728  2.2672701 -4.4405368
#> [19]  2.0089850  3.2198488
#> 
#> $sample.mean$mean
#> [1] -1.417854
#> 
#> $sample.mean$se
#> [1] 3.326092
#> 
#> 
#> $sample.se
#> $sample.se$values
#>  [1] 10.373473  8.030585  7.442725  7.024444  4.814206  7.540554 13.348165
#>  [8] 13.286942  8.528937 10.332545  7.955165 11.456520 12.740946 12.317868
#> [15]  8.362982  8.832477 12.362353  7.001484 12.401753  8.182695
#> 
#> $sample.se$mean
#> [1] 9.616841
#> 
#> $sample.se$se
#> [1] 2.519184
```

Some expected errors and warnings

``` r

## size>length(pop.2)
pop.2 <- rnorm(10,0,10)#
SRSampling(x=pop.2, size=15, rep=20) # error; not run
#> Error in SRSampling(x = pop.2, size = 15, rep = 20): size must be no greater than the length of x

## non-integer size
SRSampling(x=pop.2, size=5.8) # warning
#> Warning in SRSampling(x = pop.2, size = 5.8): size=5.8 is not an integer, and ceiling() function was applied.
#>  size=6 was used
#> A simple random sample of size 6 is drawn from x
#> 
#> The SRS for each run of samplings are NOT included in the result
#> $sample.mean
#> [1] -6.410778
#> 
#> $sample.se
#> [1] 12.14222

## x=numeric(0)
SRSampling(x=numeric(0), size=1) # error; not run
#> Error in SRSampling(x = numeric(0), size = 1): x must be a numeric vector with length >1
```

input with NA

``` r
## x with NA
pop.3 <- c(1,2,3,4,NA)
SRSampling(x=pop.3, size=5, na.rm=T)
#> Warning in SRSampling(x = pop.3, size = 5, na.rm = T): 
#> size is equal to the length of x
#> A simple random sample of size 5 is drawn from x
#> 
#> The SRS for each run of samplings are NOT included in the result
#> $sample.mean
#> [1] 2.5
#> 
#> $sample.se
#> [1] 1.290994
SRSampling(x=pop.3, size=5)
#> Warning in SRSampling(x = pop.3, size = 5): 
#> size is equal to the length of x
#> A simple random sample of size 5 is drawn from x
#> 
#> The SRS for each run of samplings are NOT included in the result
#> $sample.mean
#> [1] NA
#> 
#> $sample.se
#> [1] NA
```


NEWS.md
# SRS 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.

The first version of `SRS` is now available!

This package only contains one function `SRSampling` of taking one or multiple simple random samples (SRS) of the specified size from a assumed population which are elements of a vector with length>1, and return the summary statistics including sample mean, sample standard deviation and etc.

Take a look at the vignette to get a sense of what the function does .


tests/testthat.R
library(testthat)
library(SRS)

test_check("SRS")


tests/testthat/test-SRSampling.R
# report error when sample size > population size
test_that("exceeded size test", expect_error(SRSampling(x=rnorm(10), size=11), "size must be no greater than the length of x"))

# report error when length(x)=1
test_that("length of (x)= 1 test",expect_error(SRSampling(x=rnorm(1), size=1), "x must be a numeric vector with length >1"))

# test the length of output
rep <- 20
test_that("length of sample means", expect_length(SRSampling(x=rnorm(100), size=10, rep=20)$sample.mean$value, rep))

# test the named list output
test_that("names of the output list with show.SRS=T", expect_named(SRSampling(x=rnorm(100), size=10, rep=20, show.SRS=TRUE),c("sample.mean","sample.se", "SRS")))

test_that("names of the output list", expect_named(SRSampling(x=rnorm(100), size=10, rep=20),c("sample.mean","sample.se")))




tests/testthat/test-is_wholenumber.R
test_that("Is it a integer", {
  expect_equal(is_wholenumber(2), TRUE)
  expect_equal(is_wholenumber(2.5), FALSE)
  expect_equal(is_wholenumber(c(1,2.5,3)), c(TRUE,FALSE,TRUE))
})


R/SRSampling.R
#' SRSampling: take one or multiple simple random samples
#'
#' @param x the assumed population from which the SRS are chosen. It must be a vector with length > 1. NA in x can also be chosen.
#' @param size the sample size. the number of items to choose
#' @param replace whether sampling should be with replacement
#' @param rep the number of simulation runs. Default rep=1.
#' @param show.SRS whether to show SRS for each run of simulations
#' @param verbose whether to return verbose message
#' @param na.rm whether to remove NA when calculate summary statistics for SRS.
#' @param trim the fraction (0 to 0.5) of observations to be trimmed from each end of x before the mean is computed. Values of trim outside that range are taken as the nearest endpoint
#' @param ... additional arguments to be passed to or from methods.
#'
#' @return a list a list containing the following components, conditional on the argument `rep`:

#' When `rep=1` (the default):
#' \itemize{
#' \item `sample.mean`: a numeric number giving the mean of the SRS
#' \item `sample.se`: a numeric number giving the SD of the SRS
#' \item `SRS`: a matrix containing the values in the SRS with `nrow=1` and `ncol=size`. Available only if `show.SRS=TRUE`
#' }

#'  When `rep>1`:
#'  \itemize{
#'  \item `sample.mean`: a list containing the following components about sample means from multiple SRS:
#'    \itemize{
#'    \item `values`: a vector with a length of `rep` giving sample mean for each of the multiple SRS
#'    \item `mean`: a numeric number giving the mean of `values`
#'    \item `se`: a numeric number giving the mean of `values`
#'    }
#'  \item `sample.se`: a list containing the following components about sample sds from multiple SRS:
#'    \itemize{
#'    \item `values`: a vector with a length of `rep` giving sample sd for each of the multiple SRS
#'    \item `mean`: a numeric number giving the mean of `values`
#'    \item `se`: a numeric number giving the mean of `values`
#'    }
#'  \item `SRS`: a matrix containing the values in each of the multiple SRS with `nrow=rep` and `ncol=size`. Available only if `show.SRS=TRUE`.
#'  }

#' @export
#'
#' @examples
#' pop.1 <- rnorm(1000,0,10)
#' # take a SRS from x of size=10, show.SRS=TRUE
#' SRSampling(x=pop.1, size=10, show.SRS=TRUE)
#' # take 20 SRS from x of size=10, show.SRS=FALSE
#' SRSampling(x=pop.1, size=10, rep=20)
#'
#' # size>length(pop.2)
#' pop.2 <- rnorm(10,0,10)
#' # SRSampling(x=pop.2, size=15, rep=20) # error
#'
#' # non-integer size
#' SRSampling(x=pop.2, size=5.8) # warning
#'
#' # x=numeric(0)
#' # SRSampling(x=numeric(0), size=1) # error
#'
#' # x with NA
#' pop.3 <- c(1,2,3,4,NA)
#' SRSampling(x=pop.3, size=5, na.rm=TRUE)
#' SRSampling(x=pop.3, size=5)


SRSampling <- function(x, size, replace=FALSE, rep=1, show.SRS=FALSE,
                       verbose=TRUE, na.rm=FALSE, trim=0,...){

  # set constraints for the input arguments

  if(length(x)<=1|!is.numeric(x))
    stop("x must be a numeric vector with length >1")

  if(size<=0|!is.numeric(size)|length(size)>1)
    stop("size must be a single positive numeric argument")
  else if(!is_wholenumber(size)){
    warning(paste("size=",size," is not an integer, and ceiling() function was applied.\n size=", ceiling(size), " was used\n", sep=""))
    size=ceiling(size)
  }

  if(size>length(x))
    stop("size must be no greater than the length of x")
  if(size==length(x))
    warning("\nsize is equal to the length of x\n")

  if(rep<=0|!is.numeric(rep)|length(rep)>1)
    stop("rep must be a single positive numeric argument")
  else if(!is_wholenumber(rep)){
    warning(paste("rep=",rep," is not an integer, and ceiling() function was applied.\n rep=", ceiling(rep), " was used\n", sep=""))
    rep=ceiling(rep)
  }

  ## if rep=1; only a single SRS is taken ##
  ## return SRS(optional), sample mean, sample SD ##
  if(rep==1){
    if(verbose) cat("A simple random sample of size",size,"is drawn from x\n")
    samp <- matrix(sample(x=x, size=size, replace=replace), nrow=1)
    sample.mean<- mean(samp, trim=trim, na.rm=na.rm, ...)
    sample.se <- stats::sd(samp, na.rm=na.rm, ...)
  }

  ## if rep>1; multiple SRS is taken ##
  ## return SRS(optional),
  ## a vector of sample means; mean and SD for these sample means ##
  ## a vector of sample SDs; mean and SD for these sample SDs ##
  if(rep>1){
    if(verbose) cat(rep,"runs of simple random sample of size",size,"are drawn from x\n")
    samp <- matrix(NA,ncol=size, nrow=rep)
    samp.mean <- c()
    samp.se <- c()
    for(i in 1:rep){
      # if(verbose==TRUE) cat("\nThe ",i,"-th run of simulation\n", sep="")
      temp <- sample(x=x, size=size, replace=replace)
      temp.mean<- mean(temp, trim=trim, na.rm=na.rm, ...)
      temp.se <- stats::sd(temp, na.rm=na.rm, ...)
      samp[i,] <- temp
      samp.mean <- c(samp.mean, temp.mean)
      samp.se <- c(samp.se, temp.se)
    }
    mean.samp.mean <- mean(samp.mean, trim=trim, na.rm=na.rm, ...)
    se.samp.mean <- stats::sd(samp.mean, na.rm=na.rm)
    mean.samp.se <- mean(samp.se, trim=trim, na.rm=na.rm, ...)
    se.samp.se <- stats::sd(samp.se, na.rm=na.rm)

    sample.mean <- list(values=samp.mean, mean=mean.samp.mean, se=se.samp.mean)
    sample.se <- list(values=samp.se, mean=mean.samp.se, se=se.samp.se)
  }
  if(show.SRS==TRUE){
    if(verbose) cat("\nThe SRS for each run of samplings are shown in the result\n\n")
    result<- list(sample.mean=sample.mean, sample.se=sample.se, SRS=samp)
  }
  if(show.SRS==FALSE){
    if(verbose) cat("\nThe SRS for each run of samplings are NOT included in the result\n\n")
    result<- list(sample.mean=sample.mean, sample.se=sample.se)
  }
  return(result)
}
